python/tank/context.py

```diff
@@ -761,12 +754,73 @@ def deserialize(cls, context_str):
 
         # create a Sgtk API instance.
         tk = Tank(pipeline_config_path)
+        data["tk"] = tk
 
         # add it to the constructor instance
+        # and lastly make the obejct
+        return cls._from_dict(data)
+
+    def to_dict(self):
+        """
+        Converts the context into a dictionary with keys ``project``,
+            ``entity``, ``user``, ``step``, ``task``, ``additional_entities`` and
+            ``source_entity``.
+
+        ..note::
+            Contrary to :meth:`Context.serialize`, this method discards information
+            about the Toolkit instance associated with the context or the currently
+            authenticated user.
+
+        :returns: A dictionary representing the context.
+        """
+        return {
+            "project": self.project,
+            "entity": self.entity,
+            "user": self.user,
+            "step": self.step,
+            "task": self.task,
+            "additional_entities": self.additional_entities,
+            "source_entity": self.source_entity
+        }
+
+    @classmethod
+    def from_dict(cls, tk, data):
+        """
+        Converts a dictionary into a :class:`Context` object.
+
+        You should only pass in a dictionary that was created with the :meth:`Context.to_dict`
+        method.
+
+        :param dict data: A dictionary generated from :meth:`Context.to_dict`.
+        :param tk: Toolkit instance to associate with the context.
+        :type tk: :class:`Sgtk`
+
+        :returns: A newly created :class:`Context` object.
+        """
+        data = copy.deepcopy(data)
         data["tk"] = tk
+        return cls._from_dict(data)
 
-        # and lastly make the obejct
-        return cls(**data)
+    @classmethod
+    def _from_dict(cls, data):
+        """
+        Creates a Context object based on the arguments found in a dictionary, but
+        only the ones that the Context understands.
+
+        This ensures that if a more recent version of Toolkit serializes a context
+        and this API reads it that it won't blow-up.
+
+        :param dict data: Data for the context.
+
+        :returns: :class:`Context`
+        """
+        # Get all argument names except for self.
+        argument_names = inspect.getargspec(Context.__init__).args[1:]
```

Comment:
this feels quite magical. the `to_dict` is very explicit - any reason not to do the same here? I think that may be easier code to maintain.